### Import necessary libraries

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

2. DATA EXPLORATION (KHÁM PHÁ DỮ LIỆU - ĐAN XEN TIỀN XỬ LÝ DỮ LIỆU)

Get data in file "inf.csv" and save it into dataframe df_info

In [2]:
df_info = pd.read_csv('inf.csv')
df_info.head()

Id                                              Title  \
0  2127880  Phòng nội thất cao cấp ưu tiên nữ, gần vòng xo...   
1  2127970  Nhà Cho Thuê RỘNG RÃI - THOÁNG MÁT - AN NINH -...   
2  2127186  Cho thuê mặt bằng KD và phòng trọ số 36/66 Ngu...   
3  2122158  Thuê phòng 2.5 triệu 20m2 căn bìa tầng trệt gầ...   
4  2119311  Cho thuê phòng ĐẸP 16m2 đường Thành Thái-Q10, ...   

              Price  Square    District         City             Date  
0  2.9 Triệu/tháng     22m²     Tân Phú  Hồ Chí Minh  \r\n08/12/2022   
1  3.5 Triệu/tháng   52.5m²      Quận 9  Hồ Chí Minh  \r\n08/12/2022   
2    2 Triệu/tháng      6m²  Bình Thạnh  Hồ Chí Minh  \r\n07/12/2022   
3  2.5 Triệu/tháng     20m²      Quận 4  Hồ Chí Minh  \r\n29/11/2022   
4    5 Triệu/tháng     16m²     Quận 10  Hồ Chí Minh  \r\n25/11/2022

In [3]:
df_info = df_info.set_index('Id').sort_index()
df_info.head()

Title             Price  \
Id                                                                            
478324  Cho thuê phòng để ở tại 108/64A1 Trần Quang Di...  3.5 Triệu/tháng    
728993  Phòng vừa xây xong thoáng mát gần các trường Đ...   3.3 Triệu/tháng   
781296  Cần share lại gấp phòng trọ mới thuê tại đường...    3 Triệu/tháng    
781448  Chính chủ cho thuê căn hộ dịch vụ gần sân bay ...    7 Triệu/tháng    
789361  Phòng cho thuê giá rẻ, Quận Gò Vấp gần chung c...  1.7 Triệu/tháng    

       Square  District         City             Date  
Id                                                     
478324   18m²    Quận 3  Hồ Chí Minh  \r\n20/06/2020   
728993  25 m²    Quận 7  Hồ Chí Minh  \r\n17/10/2018   
781296   15m²   Quận 10  Hồ Chí Minh  \r\n05/10/2018   
781448   27m²  Tân Bình  Hồ Chí Minh  \r\n06/10/2018   
789361   12m²    Gò Vấp  Hồ Chí Minh  \r\n09/10/2018

In [4]:
df_info.shape

(1185, 6)

Dataset có 1185 dòng và 6 cột

Mỗi dòng cho biết thông tin của nhà trọ hoặc phòng trọ cho thuê. Có vẻ không có vấn đề các dòng có ý nghĩa khác nhau.

In [5]:
df_info.duplicated().sum()

0

Mỗi cột có ý nghĩa gì?
Ý nghĩa của mỗi cột là:

ID là để phân biệt phòng trọ, nhà trọ.

Title là tiêu đề của phòng trọ, nhà trọ.

Price là giá cho thuê của phòng trọ, nhà trọ theo tháng, đơn vị là VNĐ.

Square là diện tích của phòng trọ, nhà trọ, đơn vị là m2

District là địa chỉ quận của phòng trọ, nhà trọ.

City là địa chỉ thành phố của phòng trọ, nhà trọ.

Date là ngày đăng tin cho thuê của phòng trọ, nhà trọ.

In [6]:
df_info.columns

Index(['Title', 'Price', 'Square', 'District', 'City', 'Date'], dtype='object')

------------

KIỂM TRA KIỂU DỮ LIỆU

In [7]:
df_info.dtypes

Title       object
Price       object
Square      object
District    object
City        object
Date        object
dtype: object

Kết quả trên cho thấy tất cả các cột đều có kiểu dữ liệu là object. Gây khó khăn cho quá trình tính toán.

Ta thấy rằng có 3 cột có thể chuyển sang dạng khác là Price (numeric), Square (numeric) và Date (datetime). Tiến hành chuyển thôi nàoo

-----

In [8]:
df_info['Square']

Id
478324       18m²
728993      25 m²
781296       15m²
781448       27m²
789361       12m²
            ...  
2119311      16m²
2122158      20m²
2127186       6m²
2127880      22m²
2127970    52.5m²
Name: Square, Length: 1185, dtype: object

Ta thấy các dữ liệu để có m² ở sau, ta loại bỏ và chuyển nó sang dạng float.

In [9]:
df_info['Square'] = df_info['Square'].str.replace("m²","").to_list()
df_info['Square']= pd.to_numeric(df_info['Square'], errors='coerce')

In [10]:
df_info.rename(columns={'Square': 'Square(m²)'}, inplace = True)

In [11]:
df_info

Title             Price  \
Id                                                                             
478324   Cho thuê phòng để ở tại 108/64A1 Trần Quang Di...  3.5 Triệu/tháng    
728993   Phòng vừa xây xong thoáng mát gần các trường Đ...   3.3 Triệu/tháng   
781296   Cần share lại gấp phòng trọ mới thuê tại đường...    3 Triệu/tháng    
781448   Chính chủ cho thuê căn hộ dịch vụ gần sân bay ...    7 Triệu/tháng    
789361   Phòng cho thuê giá rẻ, Quận Gò Vấp gần chung c...  1.7 Triệu/tháng    
...                                                    ...               ...   
2119311  Cho thuê phòng ĐẸP 16m2 đường Thành Thái-Q10, ...    5 Triệu/tháng    
2122158  Thuê phòng 2.5 triệu 20m2 căn bìa tầng trệt gầ...  2.5 Triệu/tháng    
2127186  Cho thuê mặt bằng KD và phòng trọ số 36/66 Ngu...    2 Triệu/tháng    
2127880  Phòng nội thất cao cấp ưu tiên nữ, gần vòng xo...  2.9 Triệu/tháng    
2127970  Nhà Cho Thuê RỘNG RÃI - THOÁNG MÁT - AN NINH -...  3.5 Triệu/tháng    

         Square(m²)    District         City             Date  
Id                                                             
478324         18.0      Quận 3  Hồ Chí Minh  \r\n20/06/2020   
728993         25.0      Quận 7  Hồ Chí Minh  \r\n17/10/2018   
781296         15.0     Quận 10  Hồ Chí Minh  \r\n05/10/2018   
781448         27.0    Tân Bình  Hồ Chí Minh  \r\n06/10/2018   
789361         12.0      Gò Vấp  Hồ Chí Minh  \r\n09/10/2018   
...             ...         ...          ...              ...  
2119311        16.0     Quận 10  Hồ Chí Minh  \r\n25/11/2022   
2122158        20.0      Quận 4  Hồ Chí Minh  \r\n29/11/2022   
2127186         6.0  Bình Thạnh  Hồ Chí Minh  \r\n07/12/2022   
2127880        22.0     Tân Phú  Hồ Chí Minh  \r\n08/12/2022   
2127970        52.5      Quận 9  Hồ Chí Minh  \r\n08/12/2022   

[1185 rows x 6 columns]

-------

In [12]:
df_info['Price']

Id
478324     3.5 Triệu/tháng 
728993      3.3 Triệu/tháng
781296       3 Triệu/tháng 
781448       7 Triệu/tháng 
789361     1.7 Triệu/tháng 
                 ...       
2119311      5 Triệu/tháng 
2122158    2.5 Triệu/tháng 
2127186      2 Triệu/tháng 
2127880    2.9 Triệu/tháng 
2127970    3.5 Triệu/tháng 
Name: Price, Length: 1185, dtype: object

Ta thấy các dữ liệu đều có đuôi là Triệu/tháng. Còn đuôi nào nữa không nhỉ? Kiểm tra thử nhé.

In [13]:
df_info['Price'].unique()

array(['3.5 Triệu/tháng ', '3.3 Triệu/tháng', '3 Triệu/tháng ',
       '7 Triệu/tháng ', '1.7 Triệu/tháng ', '1.5 Triệu/tháng ',
       '5 Triệu/tháng', 'Thương lượng', '6 Triệu/tháng',
       '4.5 Trăm nghìn/tháng', '4 Triệu/tháng', '1.2 Triệu/tháng',
       '5.5 Triệu/tháng', '4.5 Triệu/tháng ', '2.6 Triệu/tháng ',
       '2.5 Triệu/tháng ', '3500 Triệu/tháng ', '4.7 Triệu/tháng',
       '2.3 Triệu/tháng', '1.1 Triệu/tháng', '1 Triệu/tháng',
       '3.7 Triệu/tháng', '3.5 Triệu/tháng', '2.6 Triệu/tháng',
       '7.5 Triệu/tháng', '5.8 Triệu/tháng', '3 Triệu/tháng',
       '5.9 Triệu/tháng', '2.8 Triệu/tháng', '1.3 Triệu/tháng',
       '1.8 Triệu/tháng ', '4 Triệu/tháng ', '2.1 Triệu/tháng',
       '2.5 Triệu/tháng', '1.5 Triệu/tháng', '1.9 Triệu/tháng',
       '2.2 Triệu/tháng', '3.2 Triệu/tháng', '1.25 Triệu/tháng',
       '6.8 Triệu/tháng', '2.7 Triệu/tháng ', '4.5 Triệu/tháng',
       '6.5 Triệu/tháng', '2 Triệu/tháng ', '9 Trăm nghìn/tháng',
       '2 Triệu/tháng', '2.25 Triệu/th

Các kiểu dữ liệu có đuôi là Trăm nghìn/tháng, Triệu/tháng

Và kiểu dữ liệu là Thương lượng

Ta sẽ chuyển các dữ liệu về dạng float và tiêu chuẩn là Triệu/tháng.
Thương lượng sẽ để là NaN

In [14]:
def convert_price(price):
   if 'Triệu/tháng' in price:
      price = price.replace('Triệu/tháng', '')
      return float(price)
   if 'Thương lượng' in price:
      return np.nan
   if 'Trăm nghìn/tháng' in price:
      price = price.replace('Trăm nghìn/tháng', '')
      return float(price)/10
   return np.nan

In [15]:
df_price = df_info['Price'].to_list()

In [16]:
for i in range(len(df_price)):
   df_price[i] = convert_price(df_price[i])

df_price = df_info['Price'].to_list() # convert to array

df_price = df_info['Price'].to_numpy() # convert to ndarray

In NumPy, the ndarray (n-dimensional array) is an object that stores a collection of items, which are typically numbers. By default, the ndarray is created with the "read-only" property, which means that the data stored in the array cannot be modified. 

In [17]:
print(df_price)

[3.5, 3.3, 3.0, 7.0, 1.7, 1.5, 5.0, 5.0, 5.0, nan, 6.0, 0.45, 4.0, 5.0, 1.2, 5.5, 4.5, 2.6, 4.5, 2.5, 3500.0, 0.45, 4.7, 5.5, 0.45, 0.45, 0.45, 0.45, 2.3, 5.0, 1.1, 3.3, nan, 4.0, 4.0, 2.3, 1.0, 6.0, 0.45, 3.7, 3.5, 0.45, 0.45, 2.6, 7.5, 5.8, 3.0, 0.45, 2.6, 6.0, 3.5, 5.9, 5.0, nan, 2.8, 3.5, nan, 1.3, 5.0, 1.8, 4.0, 1.7, 2.8, 2.1, 2.5, 2.3, nan, 0.45, 5.5, 2.6, 0.45, 1.5, 1.9, 2.2, 3.2, 2.5, 1.25, 6.8, 2.7, 4.5, 4.5, 6.5, 4.5, 4.5, 4.5, 2.0, 4.0, 0.9, 2.0, 3.0, 3.5, 4.0, 2.25, 3.2, 3.4, 3.4, 5.0, 3.5, 3.5, 3.7, 2.8, 3.0, 1.1, 1.6, 1.2, 1.0, 6.5, 2.8, 4.0, 2.2, 4.0, 10.0, 4.0, nan, 5.0, 1.8, 4.0, 4.0, nan, 3.5, 4.0, 5.0, 5.0, 5.0, 5.0, 2.2, 1.8, 5.0, 4.9, 7.5, 3.2, 2.0, 5.8, 3.5, 0.45, 1.7, 0.45, 0.45, 0.45, 5.3, 5.0, 5.0, 2.7, 4.0, 6.0, 3.5, 5.0, 6.2, 0.45, 5.0, 4.2, 1.5, 4.2, 4.5, 4.5, 5.0, 3.0, 6.1, 1.5, nan, 4.8, 4.5, 1.4, 4.8, 5.3, 4.8, 3.0, 4.5, 3.0, 4.5, 4.0, 2.7, 9.0, 3.5, 3.5, 3.3, 4.5, 7.0, 6.6, 4.5, 0.45, 6.5, 5.3, 4.0, 1.0, 4.0, 5.5, 6.0, 4.0, 4.5, 3.5, 8.5, 11.0, 7.0, 4.5,

Có thể thấy một số outlier ở phía trên

------

In [18]:
def convert_price(price,square):
    if 'Triệu/tháng' in price:
        price = price.replace('Triệu/tháng','')
        return float(price)
    if 'Trăm nghìn/tháng' in price:
        price = price.replace('Trăm nghìn/tháng','')
        return float(price)/10
    if 'Triệu/m2/tháng' in price:
        if pd.isna(square):
            return np.nan
        price = price.replace('Triệu/m2/tháng','')
        return float(price)*square
    if 'Trăm nghìn/m2/tháng' in price:
        if pd.isna(square):
            return np.nan
        price = price.replace('Trăm nghìn/m2/tháng','')
        price = float(price)/10
        return price*square
    if 'Nghìn/m2/tháng' in price:
        if pd.isna(square):
            return np.nan
        price = price.replace('Nghìn/m2/tháng','')
        price = float(price)/1000
        return price*square
    return np.nan


df_price_square = df_info[['Price','Square(m²)']]

In [19]:
print(df_price_square['Price'])

Id
478324     3.5 Triệu/tháng 
728993      3.3 Triệu/tháng
781296       3 Triệu/tháng 
781448       7 Triệu/tháng 
789361     1.7 Triệu/tháng 
                 ...       
2119311      5 Triệu/tháng 
2122158    2.5 Triệu/tháng 
2127186      2 Triệu/tháng 
2127880    2.9 Triệu/tháng 
2127970    3.5 Triệu/tháng 
Name: Price, Length: 1185, dtype: object


In [20]:
# for i in range(len(df_price_square)):
for i in df_price_square.index:
   df_price_square.at[i, 'Price'] = convert_price(df_price_square.loc[i]['Price'],df_price_square.loc[i]['Square(m²)'])

In [21]:
type(pd.to_numeric(df_price_square['Price']))

pandas.core.series.Series

In [22]:
df_info['Price']=df_price_square['Price'].to_list()

In [23]:
df_info.rename(columns={'Price':'Price(Triệu/tháng)'}, inplace=True)

In [24]:
df_info

Title  \
Id                                                           
478324   Cho thuê phòng để ở tại 108/64A1 Trần Quang Di...   
728993   Phòng vừa xây xong thoáng mát gần các trường Đ...   
781296   Cần share lại gấp phòng trọ mới thuê tại đường...   
781448   Chính chủ cho thuê căn hộ dịch vụ gần sân bay ...   
789361   Phòng cho thuê giá rẻ, Quận Gò Vấp gần chung c...   
...                                                    ...   
2119311  Cho thuê phòng ĐẸP 16m2 đường Thành Thái-Q10, ...   
2122158  Thuê phòng 2.5 triệu 20m2 căn bìa tầng trệt gầ...   
2127186  Cho thuê mặt bằng KD và phòng trọ số 36/66 Ngu...   
2127880  Phòng nội thất cao cấp ưu tiên nữ, gần vòng xo...   
2127970  Nhà Cho Thuê RỘNG RÃI - THOÁNG MÁT - AN NINH -...   

         Price(Triệu/tháng)  Square(m²)    District         City  \
Id                                                                 
478324                  3.5        18.0      Quận 3  Hồ Chí Minh   
728993                  3.3        25.0      Quận 7  Hồ Chí Minh   
781296                  3.0        15.0     Quận 10  Hồ Chí Minh   
781448                  7.0        27.0    Tân Bình  Hồ Chí Minh   
789361                  1.7        12.0      Gò Vấp  Hồ Chí Minh   
...                     ...         ...         ...          ...   
2119311                 5.0        16.0     Quận 10  Hồ Chí Minh   
2122158                 2.5        20.0      Quận 4  Hồ Chí Minh   
2127186                 2.0         6.0  Bình Thạnh  Hồ Chí Minh   
2127880                 2.9        22.0     Tân Phú  Hồ Chí Minh   
2127970                 3.5        52.5      Quận 9  Hồ Chí Minh   

                    Date  
Id                        
478324   \r\n20/06/2020   
728993   \r\n17/10/2018   
781296   \r\n05/10/2018   
781448   \r\n06/10/2018   
789361   \r\n09/10/2018   
...                  ...  
2119311  \r\n25/11/2022   
2122158  \r\n29/11/2022   
2127186  \r\n07/12/2022   
2127880  \r\n08/12/2022   
2127970  \r\n08/12/2022   

[1185 rows x 6 columns]

df_info.rename()

----------

In [25]:
df_info['Date']

Id
478324     \r\n20/06/2020 
728993     \r\n17/10/2018 
781296     \r\n05/10/2018 
781448     \r\n06/10/2018 
789361     \r\n09/10/2018 
                ...       
2119311    \r\n25/11/2022 
2122158    \r\n29/11/2022 
2127186    \r\n07/12/2022 
2127880    \r\n08/12/2022 
2127970    \r\n08/12/2022 
Name: Date, Length: 1185, dtype: object

Ta thấy dữ liệu dư ký tự '\n' và '\r'. Xóa bỏ chúng và đưa về dạng dữ liệu datetime.

In [26]:
df_info['Date']=df_info['Date'].str.replace('\r\n','')
df_info['Date']=df_info['Date'].str.replace(' ','')
df_info['Date']=pd.to_datetime(df_info['Date'],format = "%d/%m/%Y",errors = 'raise')

In [28]:
print(df_info.loc[478324, 'Date'])

2020-06-20 00:00:00
